apps/demos/Demos/Chat/MessageEditing/ReactJs/App.js

```diff
@@ -0,0 +1,133 @@
+import Guid from 'devextreme/core/guid';
+import {
+  currentUser,
+  messages as initialMessages,
+  allowEditingLabel,
+  allowDeletingLabel,
+  editingOptions,
+} from './data.js';
+
+const editingStrategy = {
+  enabled: true,
+  disabled: false,
+  custom: ({ component, message }) => {
+    const { items, user } = component.option();
+    const userId = user.id;
+    const lastNotDeletedMessage = items.findLast((item) => item.author?.id === userId && !item.isDeleted);
+    return message.id === lastNotDeletedMessage?.id;
+  },
+};
+export default function App() {
+  const [chatEditing, setChatEditing] = useState({
+    allowUpdating: true,
+    allowDeleting: true,
+  });
+  const store = [...initialMessages];
+  const customStore = new CustomStore({
+    key: 'id',
+    load: () =>
+      new Promise((resolve) => {
+        setTimeout(() => {
+          resolve([...store]);
+        }, 0);
+      }),
+    insert: (message) =>
+      new Promise((resolve) => {
+        setTimeout(() => {
+          store.push(message);
+          resolve(message);
+        });
+      }),
+  });
+  const dataSource = useMemo(
+    () =>
+      new DataSource({
+        store: customStore,
+        paginate: false,
+      }),
+    [],
+  );
+  const onMessageEntered = useCallback(({ message }) => {
+    const newMessage = {
+      id: new Guid().toString(),
+      ...message,
+    };
+    dataSource.store().push([
+      {
+        type: 'insert',
+        key: newMessage.id,
+        data: newMessage,
+      },
+    ]);
+  }, []);
+  const onMessageDeleted = useCallback(({ message }) => {
+    dataSource.store().push([
+      {
+        type: 'update',
+        key: message.id,
+        data: { isDeleted: true },
+      },
+    ]);
+  }, []);
+  const onMessageUpdated = useCallback(({ message, text }) => {
+    dataSource.store().push([
+      {
+        type: 'update',
+        key: message.id,
+        data: { text, isEdited: true },
+      },
+    ]);
+  }, []);
+  const handleEditingChange = useCallback((e, type) => {
+    setChatEditing({
+      [type]: editingStrategy[e.value],
+    });
+  }, []);
+  return (
+    <React.Fragment>
+      <div className="chat-container">
+        <Chat
+          height={600}
+          dataSource={dataSource}
+          user={currentUser}
+          reloadOnChange={false}
+          editing={chatEditing}
+          onMessageEntered={onMessageEntered}
+          onMessageDeleted={onMessageDeleted}
+          onMessageUpdated={onMessageUpdated}
+        />
+      </div>
+
+      <div className="options">
+        <div className="caption">Options</div>
+        <div className="option">
+          <span>Allow Editing:</span>
+          <SelectBox
+            items={editingOptions}
+            valueExpr="key"
+            displayExpr="text"
+            inputAttr={allowEditingLabel}
+            defaultValue={editingOptions[0].key}
+            onValueChanged={(e) => handleEditingChange(e, 'allowUpdating')}
```

Comment:
it will be created on each render, suggestion
```suggestion
const handleEditingChange = useCallback((type) => (e) => {
    setChatEditing({
      [type]: editingStrategy[e.value],
    });
  }, []);

...
            onValueChanged={handleEditingChange('allowUpdating')}
```

Reply:
refactored